File: year2/systems-design-implementation/Lab3/Lab3/src/main/java/repo/database/InscriereRepositoryDB.java
package repo.database;

import domain.Inscriere;
import domain.Inscriere;
import domain.Participant;
import domain.Proba;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repo.interfaces.InscriereRepository;
import utils.Pair;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class InscriereRepositoryDB implements InscriereRepository<Pair<Integer, Integer>, Inscriere> {
    private final static Logger logger = LogManager.getLogger(InscriereRepositoryDB.class);
    private String url;

    public InscriereRepositoryDB(String url) {
        this.url = url;
    }

    @Override
    public Optional<Inscriere> findOne(Pair<Integer, Integer> id) throws SQLException {
        logger.traceEntry();
        Optional<Inscriere> i = Optional.empty();
        Connection connection = DriverManager.getConnection(url);
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Inscrieri where id_participant=? and id_proba=?")) {
            preparedStatement.setInt(1, id.getE1());
            preparedStatement.setInt(2, id.getE2());
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                i = Optional.of(new Inscriere(id));
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(i.get());
        return i;
    }

    @Override
    public Iterable<Inscriere> findAll() throws SQLException {
        logger.traceEntry();
        Connection connection = DriverManager.getConnection(url);
        List<Inscriere> list = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Inscrieri")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    int id_participant = resultSet.getInt("id_participant");
                    int id_proba = resultSet.getInt("id_proba");
                    Inscriere i = new Inscriere(new Pair<>(id_participant, id_proba));
                    list.add(i);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(list);
        return list;
    }

    @Override
    public Optional<Inscriere> add(Inscriere entity) throws SQLException {
        logger.traceEntry("Adding task {} ", entity);
        Optional<Inscriere> i = Optional.of(entity);
        Connection connection = DriverManager.getConnection(url);
        try (PreparedStatement preStmt = connection.prepareStatement("insert into Inscrieri (id_participant, id_proba) values (?,?)")) {
            preStmt.setInt(1, entity.getId().getE1());
            preStmt.setInt(2, entity.getId().getE2());
            int result = preStmt.executeUpdate();
            logger.trace("Added {} instance", result);
            i = Optional.empty();
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit();
        return i;
    }

    @Override
    public Optional<Inscriere> delete(Pair<Integer, Integer> id) throws SQLException {
        logger.traceEntry();
        Optional<Inscriere> i = findOne(id);
        Connection connection = DriverManager.getConnection(url);
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from Inscrieri where id_participant=? and id_proba=?")) {
            preparedStatement.setInt(1, id.getE1());
            preparedStatement.setInt(2, id.getE2());
            preparedStatement.executeUpdate();
            logger.trace("Removed {} instance", i);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit();
        return i;
    }

    @Override
    public Optional<Inscriere> update(Inscriere entity) throws SQLException {
        logger.traceEntry();
        Optional<Inscriere> i = findOne(entity.getId());
        Connection connection = DriverManager.getConnection(url);
        try (PreparedStatement preparedStatement = connection.prepareStatement("update Inscrieri set id_participant=?,id_proba=? where id_participant=? and id_proba=?")) {
            preparedStatement.setInt(1, entity.getId().getE1());
            preparedStatement.setInt(2, entity.getId().getE2());
            preparedStatement.setInt(3, entity.getId().getE1());
            preparedStatement.setInt(4, entity.getId().getE2());
            preparedStatement.executeUpdate();
            logger.trace("Updated {} instance", i);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit();
        return i;
    }

    @Override
    public Iterable<Participant> getInscrisi(Proba proba) throws SQLException {
        logger.traceEntry();
        Connection connection = DriverManager.getConnection(url);
        List<Participant> list = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select P.id, P.nume, P.varsta from Inscrieri I inner join Participanti P on I.id_participant=P.id where I.id_proba=?")) {
            preparedStatement.setInt(1, proba.getId());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    int id = resultSet.getInt("id");
                    String nume = resultSet.getString("nume");
                    int varsta = resultSet.getInt("varsta");
                    Participant p = new Participant(id, nume, varsta);
                    list.add(p);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(list);
        return list;
    }

    @Override
    public Integer nrParticipanti(Proba proba) throws SQLException {
        logger.traceEntry();
        int nr = 0;
        Connection connection = DriverManager.getConnection(url);
        try (PreparedStatement preparedStatement = connection.prepareStatement("select count(P.id) as nr from Inscrieri I inner join Participanti P on I.id_participant=P.id where I.id_proba=?")) {
            preparedStatement.setInt(1, proba.getId());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    nr = resultSet.getInt("nr");
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(nr);
        return nr;
    }
}

File: year2/systems-design-implementation/Lab3/Lab3Sharp/Repo/Database/InscriereRepositoryDB.cs
using Lab2Sharp.Domain;
using Lab2Sharp.Repo.Interfaces;
using Lab2Sharp.Utils;
using log4net;
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab2Sharp.Repo.Database
{
    class InscriereRepositoryDB<ID, E> : InscriereRepository<Pair<int, int>, Inscriere>
    {
        private string url;
        private static readonly ILog log = LogManager.GetLogger("ParticipantRepository");

        public InscriereRepositoryDB(string url)
        {
            this.url = url;
        }

        public Inscriere FindOne(Pair<int, int> id)
        {
            string sqlCom = "select * from Inscrieri where id_participant=@id1 and id_proba=@id2";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id1 = cmd.CreateParameter();
                    id1.ParameterName = "@id1";
                    id1.Value = id.e1;
                    cmd.Parameters.Add(id1);
                    var id2 = cmd.CreateParameter();
                    id2.ParameterName = "@id2";
                    id2.Value = id.e2;
                    cmd.Parameters.Add(id2);
                    using (var data = cmd.ExecuteReader())
                    {
                        if (data.Read())
                        {
                            Inscriere p = new Inscriere(id.e1, id.e2);
                            return p;
                        }
                    }
                }
            }
            return null;
        }
        public IEnumerable<Inscriere> FindAll()
        {
            log.InfoFormat("Entering findAll");
            IList<Inscriere> list = new List<Inscriere>();
            string sqlCom = "select * from Inscrieri";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                var cmd = new SQLiteCommand(sqlCom, conn);
                using (var data = cmd.ExecuteReader())
                {
                    while (data.Read())
                    {
                        int id_participant = data.GetInt32(0);
                        int id_proba = data.GetInt32(1);
                        Inscriere p = new Inscriere(id_participant, id_proba);
                        log.InfoFormat("Found {0}", p);
                        list.Add(p);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return list;
        }
        public Inscriere Add(Inscriere entity)
        {
            Inscriere p = entity;
            string sqlCom = "insert into Inscrieri (id_participant, id_proba) values (@id1,@id2)";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id1 = cmd.CreateParameter();
                    id1.ParameterName = "@id1";
                    id1.Value = entity.id.e1;
                    cmd.Parameters.Add(id1);
                    var id2 = cmd.CreateParameter();
                    id2.ParameterName = "@id2";
                    id2.Value = entity.id.e2;
                    cmd.Parameters.Add(id2);
                    if (cmd.ExecuteNonQuery() > 0)
                        p = null;
                }
            }
            return p;
        }
        public Inscriere Delete(Pair<int, int> id)
        {
            Inscriere p = FindOne(id);
            string sqlCom = "delete from Inscrieri where id_participant=@id1 and id_proba=@id2";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {

                    var id1 = cmd.CreateParameter();
                    id1.ParameterName = "@id1";
                    id1.Value = id.e1;
                    cmd.Parameters.Add(id1);
                    var id2 = cmd.CreateParameter();
                    id2.ParameterName = "@id2";
                    id2.Value = id.e2;
                    cmd.Parameters.Add(id2);
                    cmd.ExecuteNonQuery();
                }
            }
            return p;
        }
        public Inscriere Update(Inscriere entity)
        {
            Inscriere p = FindOne(entity.id);
            string sqlCom = "update Inscrieri set id_participant=@id1,id_proba=@id2 where id_participant=@id1 and id_proba=@id2";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id1 = cmd.CreateParameter();
                    id1.ParameterName = "@id1";
                    id1.Value = entity.id.e1;
                    cmd.Parameters.Add(id1);
                    var id2 = cmd.CreateParameter();
                    id2.ParameterName = "@id2";
                    id2.Value = entity.id.e2;
                    cmd.Parameters.Add(id2);
                    cmd.ExecuteNonQuery();
                }
            }
            return p;
        }

        public IEnumerable<Participant> GetInscrisi(Proba proba)
        {
            IList<Participant> list = new List<Participant>();
            string sqlCom = "select P.id, P.nume, P.varsta from Inscrieri I inner join Participanti P on I.id_participant=P.id where I.id_proba=@id";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id = cmd.CreateParameter();
                    id.ParameterName = "@id";
                    id.Value = proba.id;
                    cmd.Parameters.Add(id);

                    using (var data = cmd.ExecuteReader())
                    {
                        while (data.Read())
                        {
                            int idd = data.GetInt32(0);
                            String nume = data.GetString(1);
                            int varsta = data.GetInt32(2);
                            Participant p = new Participant(idd, nume, varsta);
                            list.Add(p);
                        }
                    }
                }
            }

            return list;
        }
        public int nrParticipanti(Proba proba)
        {
            string sqlCom = "select count(P.id) as nr from Inscrieri I inner join Participanti P on I.id_participant=P.id where I.id_proba=@id";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id = cmd.CreateParameter();
                    id.ParameterName = "@id";
                    id.Value = proba.id;
                    cmd.Parameters.Add(id);
                    using (var data = cmd.ExecuteReader())
                    {
                        if (data.Read())
                        {
                            return data.GetInt32(0);
                        }
                    }
                }
            }
            return 0;
        }
    }
}
